### PR TITLE
DEC-152: User roles and permissions

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Adds new default users to match the new default roles

# Review By (Date)
- EOD Tuesday October 15

# Urgency
- Medium, we'd like to get some of this out to the pilot.

# Steps to Test

1. Read and review https://docs.google.com/document/d/1X3lqcf7m1awvF6J8PkxdfUy-ZAiBklx9GUKMf2LOvxA/edit
2. Check out this branch
3. Run full site install
4. Log in as an administrator
5. Validate users and roles are there with no typos
6. Review permissions page and compare options to the google document
7. Run tests

# Affected Projects or Products
- This profile

# Associated Issues and/or People
- https://docs.google.com/document/d/1X3lqcf7m1awvF6J8PkxdfUy-ZAiBklx9GUKMf2LOvxA/edit
- D8CORE-152
- https://github.com/SU-SWS/acsf-cardinalsites/pull/27

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: config/sync/user.role.site_builder.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+id: site_builder
+label: 'Site Builder'
+weight: -1
+is_admin: null
+permissions:
+  - 'access administration pages'
+  - 'access config_pages overview'
+  - 'access content overview'
+  - 'access contextual links'
+  - 'access devel information'
```
Comment (pookmish) [inline config/sync/user.role.site_builder.yml]: Probably don't want any devel in any of these permissions. the module isn't on anything other than local.

File: config/sync/user.role.site_developer.yml [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  - 'administer config_pages fields'
+  - 'administer config_pages form display'
+  - 'administer config_pages types'
+  - 'administer content types'
+  - 'administer crop'
+  - 'administer crop types'
+  - 'administer devel menu items'
+  - 'administer display modes'
+  - 'administer embed buttons'
+  - 'administer entity browsers'
```
Comment (pookmish) [inline config/sync/user.role.site_developer.yml]: This is going to be removed after stanford_media update is in place

File: config/sync/user.role.site_editor.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  - 'access content overview'
+  - 'access contextual links'
+  - 'access environment indicator'
+  - 'access file_browser entity browser pages'
+  - 'access files overview'
+  - 'access image_browser entity browser pages'
+  - 'access media overview'
+  - 'access media_browser entity browser pages'
+  - 'access patterns page'
+  - 'access video_browser entity browser pages'
```
Comment (pookmish) [inline config/sync/user.role.site_editor.yml]: the entity browser pages are going to be removed after stanford_media update is in place

File: config/sync/user.role.site_editor.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  - 'delete own stanford_page content'
+  - 'delete own video media'
+  - 'delete stanford_page revisions'
+  - 'dropzone upload files'
+  - 'edit any stanford_page content'
+  - 'edit own file media'
+  - 'edit own image media'
+  - 'edit own stanford_page content'
+  - 'edit own video media'
+  - 'perform batch updates entity usage'
```
Comment (pookmish) [inline config/sync/user.role.site_editor.yml]: all roles dont really need this. It just bulk updates usages data.

File: config/sync/user.role.site_manager.yml [hunk 5]
```diff
@@ -7,23 +7,74 @@ label: 'Site Manager'
 weight: -3
 is_admin: null
 permissions:
+  - 'access administration pages'
   - 'access content overview'
+  - 'access contextual links'
+  - 'access environment indicator'
   - 'access file_browser entity browser pages'
+  - 'access files overview'
```
Comment (pookmish) [inline config/sync/user.role.site_manager.yml]: Do we really wanna give them this page? they will have access to the media overview page

File: config/sync/user.role.site_builder.yml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  - 'administer blocks'
+  - 'administer content types'
+  - 'administer crop'
+  - 'administer crop types'
+  - 'administer display modes'
+  - 'administer entity browsers'
+  - 'administer footer menu items'
+  - 'administer image styles'
+  - 'administer main menu items'
+  - 'administer menu'
```
Comment (pookmish) [inline config/sync/user.role.site_builder.yml]: Do we really wanna give this permission? they could modify the admin menu with this permission. With the menu admin per menu module, the permissions restrict them to only main nav and footer nav

File: config/sync/user.role.site_builder.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  - 'administer taxonomy'
+  - 'administer taxonomy_term display'
+  - 'administer taxonomy_term fields'
+  - 'administer taxonomy_term form display'
+  - 'administer themes'
+  - 'administer url aliases'
+  - 'administer user display'
+  - 'administer user fields'
+  - 'administer user form display'
+  - 'administer users'
```
Comment (pookmish) [inline config/sync/user.role.site_builder.yml]: I think we need to turn on and enable role delegation too. This permission allows them to add a sunet user with any role

File: config/sync/user.role.site_editor.yml [hunk 8]
```diff
@@ -0,0 +1,53 @@
+uuid: b35dd6b6-930d-4b4c-aeb2-443a0ba6fb59
+langcode: en
+status: true
+dependencies: {  }
+id: site_editor
+label: 'Site Editor'
+weight: -2
+is_admin: null
+permissions:
```
Comment (pookmish) [inline config/sync/user.role.site_editor.yml]: - User can't unpublish any content.

File: config/sync/user.role.contributor.yml [hunk 9]
```diff
@@ -7,15 +7,32 @@ label: Contributor
 weight: -4
 is_admin: null
 permissions:
```
Comment (pookmish) [inline config/sync/user.role.contributor.yml]: - This role can only edit their own content. The google doc does not mention if this is the desire.
- They currently cant view or revert revisions
- They can delete their own content but cant unpublish
- They can edit the main menu & delete/create/rearrange links

File: config/sync/user.role.site_manager.yml [hunk 10]
```diff
@@ -7,23 +7,74 @@ label: 'Site Manager'
...
+  - 'access media overview'
   - 'access media_browser entity browser pages'
   - 'access patterns page'
+  - 'access site in maintenance mode'
+  - 'access taxonomy overview'
+  - 'access user profiles'
+  - 'access video_browser entity browser pages'
+  - 'administer footer menu items'
   - 'administer main menu items'
   - 'administer media'
```
Comment (pookmish) [inline config/sync/user.role.site_manager.yml]: - This gives access to `/admin/config/media/media-settings` path. i dont think we'd want them in there. we might have to change the permission on that specific path

File: config/sync/user.role.site_manager.yml [hunk 11]
```diff
@@ -7,23 +7,74 @@ label: 'Site Manager'
 weight: -3
 is_admin: null
 permissions:
```
Comment (pookmish) [inline config/sync/user.role.site_manager.yml]: - We'll need to add the site title and some things to the config pages entity.
- can't edit metadata at the moment

File: content/user/0ee1bcb8-f26c-4809-8f58-3efb932dbe35.json [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    },
+    "uid": [
+        {
+            "value": 5
+        }
+    ],
+    "uuid": [
+        {
+            "value": "0ee1bcb8-f26c-4809-8f58-3efb932dbe35"
```
Comment (pookmish) [inline content/user/0ee1bcb8-f26c-4809-8f58-3efb932dbe35.json]: add the new user UUID's to the stanford_profile.info.yml file for default content tracking

File: config/sync/user.role.site_developer.yml (hunk 2 above)
Comment (sherakama) [reply config/sync/user.role.site_developer.yml]: GTK, will update then.

File: config/sync/user.role.site_editor.yml (hunk 3 above)
Comment (sherakama) [reply config/sync/user.role.site_editor.yml]: TY.

File: config/sync/user.role.site_editor.yml (hunk 4 above)
Comment (sherakama) [reply config/sync/user.role.site_editor.yml]: GTK, TY.

File: config/sync/user.role.site_manager.yml (hunk 5 above)
Comment (sherakama) [reply config/sync/user.role.site_manager.yml]: The site manager is a pretty high-level role. I think this should be fine but I can remove it and grant later. Better to give than to take awayeth.

File: config/sync/user.role.site_builder.yml (hunk 6 above)
Comment (sherakama) [reply config/sync/user.role.site_builder.yml]: This role is generally targeting our producers or someone at a similar level. I think we should be ok with granting this permission for this role. At this first pass I'd rather avoid adding another contrib module.

File: config/sync/user.role.site_builder.yml (hunk 7 above)
Comment (sherakama) [reply config/sync/user.role.site_builder.yml]: 🤔 This sounds like privilege escalation and should be stopped. Thanks. I'll revise.

File: config/sync/user.role.site_builder.yml (hunk 6 above)
Comment (pookmish) [reply config/sync/user.role.site_builder.yml]: we've already added menu admin per menu module. so we could just rely on that

Comment (sherakama) [reply config/sync/user.role.site_builder.yml]: The module is already there. Ignore previous comment.

Comment (pookmish) [reply config/sync/user.role.site_builder.yml]: Ignoring is what i do best

File: config/sync/user.role.site_editor.yml (hunk 8 above)
Comment (sherakama) [reply config/sync/user.role.site_editor.yml]: Thanks, I'll add the `Administer content` permission.

**Comment by sherakama** [comment]
@pookmish Ok, thank you for the first thorough review. Please, can I have another?

**Comment by pookmish** [comment]
Looks like some failing tests. And I was actually working on a similar thing to set the site title. I was creating a more dynamic method that was less work to override a config when using the config pages. Lets huddle to review before too much more work is done here

**Comment by sherakama** [comment]
Ah, yeah, probably a good call.

This is reminding me of how we used to abuse webforms so we can do more. I'll grab ya sometime soon to compare notes.

**Comment by sherakama** [comment]
@pookmish Config pages stuff removed. Back to you.

File: config/sync/user.role.site_manager.yml [hunk 13]
```diff
@@ -7,23 +7,73 @@ label: 'Site Manager'
...
   - 'access media_browser entity browser pages'
   - 'access patterns page'
+  - 'access site in maintenance mode'
+  - 'access taxonomy overview'
+  - 'access user profiles'
+  - 'access video_browser entity browser pages'
+  - 'administer footer menu items'
   - 'administer main menu items'
-  - 'administer media'
+  - 'administer menu'
```
Comment (pookmish) [inline config/sync/user.role.site_manager.yml]: Now the site manager gets access to change the admin menu and more

File: config/sync/user.role.site_manager.yml [hunk 14]
```diff
@@ -7,23 +7,73 @@ label: 'Site Manager'
...
+  - 'edit any file media'
+  - 'edit any image media'
   - 'edit any stanford_page content'
+  - 'edit any video media'
+  - 'edit own file media'
+  - 'edit own image media'
   - 'edit own stanford_page content'
+  - 'edit own video media'
   - 'edit stanford_basic_site_settings config page entity'
+  - 'edit stanford_branding_options config page entity'
```
Comment (pookmish) [inline config/sync/user.role.site_manager.yml]: This has been removed now

File: config/sync/user.role.site_builder.yml [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  - 'administer paragraphs settings'
+  - 'administer paragraphs types'
+  - 'administer redirects'
+  - 'administer responsive images'
+  - 'administer shortcuts'
+  - 'administer taxonomy'
+  - 'administer taxonomy_term display'
+  - 'administer taxonomy_term fields'
+  - 'administer taxonomy_term form display'
+  - 'administer themes'
```
Comment (pookmish) [inline config/sync/user.role.site_builder.yml]: Do we want to have these kinda things right now? since we dont allow switching themes or config changes

Comment (sherakama) [reply config/sync/user.role.site_builder.yml]: were going to need theme switching eventually as we want our site builders to be able to push custom themes. 

As a general statement, we probably should hide the stuff that we want to keep locked so I will mull on this.

Comment (sherakama) [reply config/sync/user.role.site_builder.yml]: I'd say it is fine for Pilot and we can let the user testing provide the appropriate feedback to help make this decision.

Comment (pookmish) [reply config/sync/user.role.site_builder.yml]: sounds good

**Pull request merged**